core/capabilities/integration_tests/framework/don.go

```diff
@@ -0,0 +1,364 @@
+	"github.com/smartcontractkit/chainlink/v2/core/internal/cltest"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/chainlink"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/ethkey"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/ocr2key"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+	"github.com/smartcontractkit/chainlink/v2/core/services/standardcapabilities"
+	"github.com/smartcontractkit/chainlink/v2/core/utils/testutils/heavyweight"
+)
+
+type capabilityNode struct {
+	*cltest.TestApplication
+	registry  *capabilities.Registry
+	key       ethkey.KeyV2
+	KeyBundle ocr2key.KeyBundle
+	peerID    peer
+	start     func()
+}
+
+type DON struct {
+	t                      *testing.T
+	config                 DonConfiguration
+	lggr                   logger.Logger
+	nodes                  []*capabilityNode
+	standardCapabilityJobs []*job.Job
+	externalCapabilities   []capability
+	capabilitiesRegistry   *CapabilitiesRegistry
+
+	nodeConfigModifiers []func(c *chainlink.Config, node *capabilityNode)
+
+	addOCR3NonStandardCapability bool
+
+	triggerFactories []TriggerFactory
+	targetFactories  []TargetFactory
+}
+
+func NewDON(ctx context.Context, t *testing.T, lggr logger.Logger, donConfig DonConfiguration,
+	dependentDONs []commoncap.DON, donContext DonContext) *DON {
+	don := &DON{t: t, lggr: lggr.Named(donConfig.name), config: donConfig, capabilitiesRegistry: donContext.capabilityRegistry}
+
+	for i, member := range donConfig.Members {
+		dispatcher := donContext.p2pNetwork.NewDispatcherForNode(member)
+		capabilityRegistry := capabilities.NewRegistry(lggr)
+
+		nodeInfo := commoncap.Node{
+			PeerID:         &member,
+			WorkflowDON:    donConfig.DON,
+			CapabilityDONs: dependentDONs,
+		}
+
+		cn := &capabilityNode{
+			registry:  capabilityRegistry,
+			key:       donConfig.keys[i],
+			KeyBundle: donConfig.KeyBundles[i],
+			peerID:    donConfig.peerIDs[i],
+		}
+		don.nodes = append(don.nodes, cn)
+
+		cn.start = func() {
+			node := startNewNode(ctx, t, lggr.Named(donConfig.name+"-"+strconv.Itoa(i)), nodeInfo, donContext.EthBlockchain,
+				donContext.capabilityRegistry.getAddress(), dispatcher,
+				peerWrapper{peer: p2pPeer{member}}, capabilityRegistry,
+				donConfig.keys[i], func(c *chainlink.Config) {
+					for _, modifier := range don.nodeConfigModifiers {
+						modifier(c, cn)
+					}
+				})
+
+			require.NoError(t, node.Start(testutils.Context(t)))
+			cn.TestApplication = node
+		}
+	}
+
+	return don
+}
+
+// ExposeExternalCapabilities must be called after all capabilities have been added to the DONs and before Start is called
+func (d *DON) ExposeExternalCapabilities() {
```

Comment:
This sounds like a bit of temporal coupling: would we ever want to Start() without exposing external capabilities?

Reply:
Ideally, however the contract on which the don and capabilities are registered also determines the DON ID,  which is then required to setup the keystone contracts before the dons are started, so it had to be split into parts.  I've change the name back to Initialise as really the method is about registering the capabilities and getting the DON's ID